[CmdBuffer] Draft experimental cmd buffer YAML

Bensuo opened the pull request:
Initial draft of YAML for experimental command buffer feature.

I've only included the YAML for now since I'm not sure what generated code needs to be committed yet.

igchor committed:
[ur] Move usm_allocator implementation from llvm

From version:
https://github.com/intel/llvm/commit/edd900201a045595b818707fa90dea3926050c23

igchor committed:
[ur] refactor usm_allocator into disjoint_pool

Remove USMAllocContext and replace it with DisjointPool class with
UMA-compatible interface.

Move config parsing from USMAllocatorConfig ctor to a separate function
and add tests for that parsing.

Co-authored-by: Damian Duy <damian.duy@intel.com>

Commit:
Add kernel execution test fixture.

Also add new kernel execution tests wherever relevant.

veselypeta committed:
[UR] Refactor urDevicePartition to use desc struct

kbenzie committed:
[ur] Add experimental feature contribution docs

This is a first draft of the contribution guide for experimental
features. It does not attempt to address every eventuality but instead
present a starting point which can be built upon with time and
experience in enabling contributions.

Commit:
[ur] fix minor coverity issues

veselypeta committed:
[UR] Refactor device partition into properties struct

veselypeta committed:
[UR] Fix device partitioning programming guide

veselypeta committed:
[UR] Fix build and spec

PatKamin committed:
Fix Coverity issues

veselypeta committed:
[UR] Remove BY_COUNTS_LIST_END

wlemkows committed:
[UR] Add UI information

Commit:
Add generated kernel name header instead of relying on kernel name query.

kbenzie committed:
[ur] Fix modulo by zero Coverity defect

Replace uses of `{buffer,src,dst}RowPitch` validation logic for the
`urMemBuffer{Read,Write,Copy}Rect` entry-points with a ternary
expression which checks the value, uses it when non-zero, and use
`region.width` otherwise. This matches the OpenCL documentation for
error conditions for the analogous entry-points. While the Coverity
defect only mentions modulo by zero, this patch also updates instances
where `{buffer,src,dst}RowPitch` is used in multiplication resulting in
a buggy comparison.

PatKamin committed:
Replace logger exceptions with error messages

Wrapping every log message in a try-catch can be annoying.
Reducing exceptions throws with a message to std::cerr in cases where
messages have wrong syntax.

Commit:
[UR] Verify context device GetInfo

pbalcer committed:
Merge pull request #551 from PatKamin/logger-nothrow

[logger/coverity] Replace logger exceptions with error messages

pbalcer committed:
Merge pull request #550 from kbenzie/benie/fix-coverity-divide-by-zero

[ur] Fix modulo by zero Coverity defect

pbalcer committed:
Merge pull request #545 from wlemkows/u-id

[UR] Add UI information

Commit:
[UR] Verify event GetInfo

Commit:
[UR] Verify kernel GetInfo

Commit:
[UR] Verify memory GetInfo

Commit:
[UR] Verify platform GetInfo

Commit:
[UR] Verify program GetInfo

Commit:
[UR] Verify queue GetInfo

Commit:
[UR] Verify sampler GetInfo

Commit:
[UR] Verify USM GetInfo

Commit:
[UR] Rename UR_DEVICE_INFO_PARTITION_PROPERTIES

Commit:
[UR] Fix yaml tabs

veselypeta committed:
Merge pull request #536 from veselypeta/petr/device-partition-properties

[UR] Refactor urDevicePartition to use desc struct

EwanC commented on scripts/core/exp-command-buffer.yml:
```diff
@@ -0,0 +1,246 @@
+--- #--------------------------------------------------------------------------
+type: header
+desc: "Intel $OneApi Command Buffer Extension APIs"
```
Nitpick, but preference for consistently hyphenating "command buffer" as command-buffer

EwanC commented on scripts/core/exp-command-buffer.yml:
```diff
@@ -0,0 +1,246 @@
+members: 
+--- #--------------------------------------------------------------------------
+type: typedef
+desc: "A value that identifies a command inside of a command buffer, used for defining dependencies between commands."
```
"between commands *in the same command-buffer*"

EwanC commented on scripts/core/exp-command-buffer.yml:
```diff
@@ -0,0 +1,246 @@
+      desc: "[out] pointer to Command Buffer handle"
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Get a reference to a command buffer handle. Increment the command buffer object's reference count."
```
"Get a reference to a command buffer handle." could maybe remove this starting sentence, "get" makes me think the entry-point is returning something, which it's not doing.

EwanC commented on scripts/core/exp-command-buffer.yml:
```diff
@@ -0,0 +1,246 @@
+      desc: "[in] A list of sync points that this command depends on."
+    - type: "$x_exp_command_buffer_sync_point_t*"
+      name: pSyncPoint
+      desc: "[out] sync point associated with this kernel command"
```
Not a kernel command, same for other memory command descriptions

veselypeta committed:
Merge pull request #554 from veselypeta/petr/549/get-info-validation

Make GetInfo queries consistent and tested

Commit:
Change mit license to apache 2.0 with llvm exceptions

Commit:
Change mention of "SVM" in KernelSetArgPointer spec to "USM".

Bensuo committed:
[Experimental] Add command-buffers

- Adds support for command-buffers experimental feature
- Add command-buffer desc type to common.yml
- Add device info query for command-buffer support

Commit:
Merge pull request #559 from aarongreig/aaron/kernelPointerArgSVMtoUSM

Change mention of "SVM" in KernelSetArgPointer spec to "USM".

omarahmed1111 committed:
Merge pull request #557 from omarahmed1111/change-ur-license-to-apache

Change project license from MIT to apache 2.0 with llvm exceptions

igchor committed:
[ur] fix DisjointPool::allocate

Pointer obtained from getSlab and getChunk can be aligned
arbitrarly and might need to be adjusted by more than the
difference between Size and AlignedSize.

pbalcer committed:
Merge pull request #544 from PatKamin/coverity-issues

[coverity] Fix Coverity issues

pbalcer committed:
Merge pull request #541 from DamianDuy/fixCoverityIssues

[ur] fix minor coverity issues

martygrant committed:
Add UR_PROFILING_INFO_COMMAND_COMPLETE to ur_profling_info_t.

martygrant committed:
Merge pull request #561 from martygrant/martin/event-profiling-complete

Add UR_PROFILING_INFO_COMMAND_COMPLETE to ur_profling_info_t.

Bensuo committed:
[Experimental] Minor CmdBuffer improvements

callumfare committed:
[ur] Make hArgValue optional again in urKernelSetMemObj

callumfare committed:
Merge pull request #564 from callumfare/callum/allow_null_mem_obj_arg

[ur] Make hArgValue optional again in urKernelSetMemObj

Commit:
A number of small fixes for the kernel and program CTS tests.

Commit:
Merge pull request #553 from aarongreig/aaron/kernelExecutionTests

Add kernel execution tests.

callumfare committed:
Merge pull request #546 from kbenzie/benie/experimental-features

[ur] Add experimental feature contribution docs

Commit:
[UR] Fail build on spec generate

Commit:
Merge pull request #570 from aarongreig/aaron/kernelProgramTestFixes

A number of small fixes for the kernel and program CTS tests.

Bensuo committed:
[Experimental][CmdBuffer] Improve returns and naming

- Add return codes to all cmdbuffer functions
- Consistent naming for functions that append to a cmdbuffer

veselypeta committed:
Merge pull request #569 from veselypeta/petr/565/fail-build-on-spec-generate

[UR] Fail build on spec generate

Commit:
[UR] Documents generator and conformance tests

veselypeta committed:
Merge pull request #574 from veselypeta/petr/contrib-additions

[UR] Documents generator and conformance tests

Bensuo committed:
[Experimental][CmdBuffer] Remove device info query and associated enum

Commit:
[UR] re-raise exception in makoWrite

veselypeta committed:
Merge pull request #579 from veselypeta/petr/raise-exception-in-mako-write

[UR] re-raise exception in makoWrite

kbenzie committed:
[ur] Actually fail when RST validation finds an error

Following on from #569 and #579 with the goal of failing the build when
the generation scripts have errors, this patch introduces exceptions to
the `_generate_valid_rst` where messages containing `error :` were being
printed.

kbenzie committed:
[ur] Link to contribution guide in readme

kbenzie committed:
Merge pull request #581 from kbenzie/benie/update-readme

[ur] Link to contribution guide in readme

kbenzie committed:
Merge pull request #580 from kbenzie/benie/add-validate-rst-exceptions

[ur] Actually fail when RST validation finds an error

igchor committed:
Merge pull request #568 from igchor/usm_allocator

Move USM allocator implementation from intel/llvm

Bensuo committed:
[Experimental][CmdBuffer] Addressing PR comments

- Set command-buffer ordinal to 99
- Include license
- Fix incorrect naming conventions
- Make sync point parameters optional
- Move and add explicit values to errors

Bensuo committed:
[Experimental][CmdBuffer] Fix license

Bensuo committed:
[Experimental][CmdBuffer] Add missing colon to return errors
